plugin-common/src/main/java/org/opennms/plugins/cloud/grpc/GrpcExecutionHandler.java

```diff
@@ -44,62 +45,80 @@
+import lombok.Builder;
+import lombok.Data;
 
 /**
- * Tries to distinguish between
+ * Manages functionality for GRPC calls:
+ * - ExceptionHandling: Tries to distinguish between
  * - recoverable Exceptions: will be propgated to OpenNMS and
  * - non recoverable Exceptions: will be logged dropped
+ * - Request logging (logs are send towards the cloud)
  * see also: https://www.grpc.io/docs/guides/error/
  */
-public class GrpcExceptionHandler {
+public class GrpcExecutionHandler {
 
-    private static final Logger LOG = LoggerFactory.getLogger(GrpcExceptionHandler.class);
+    private static final Logger LOG = LoggerFactory.getLogger(GrpcExecutionHandler.class);
     private static final Set<Code> RECOVERABLE_EXCEPTIONS = new HashSet<>(Arrays.asList(
             DEADLINE_EXCEEDED,
             UNAVAILABLE,
             UNAUTHENTICATED,
             RESOURCE_EXHAUSTED));
 
-    private GrpcExceptionHandler() {
-        // Utility class
+    final CloudLogService cloudLog;
+
+    public GrpcExecutionHandler(final CloudLogService cloudLog) {
+        this.cloudLog = Objects.requireNonNull(cloudLog);
     }
 
-    public static <T, R> R executeRpcCall(Supplier<T> callToExecute, Function<T, R> mapper, Supplier<R> defaultFunction) throws StorageException {
+    public <T, R> R executeRpcCall(GrpcCall<T, R> callToExecute) throws StorageException {
+        Objects.requireNonNull(callToExecute);
+        Objects.requireNonNull(callToExecute.getMethodDescriptor());
+        Objects.requireNonNull(callToExecute.getMapper());
+        Objects.requireNonNull(callToExecute.getDefaultFunction());
+        Status.Code status = OK;
+        long startTime = System.currentTimeMillis();
         try {
-            T result = callToExecute.get();
-            return mapper.apply(result);
+            T result = callToExecute.getCallToExecute().get();
+            return callToExecute.getMapper().apply(result);
         } catch (StatusRuntimeException ex) {
-            Status.Code status = ex.getStatus().getCode();
+            status = ex.getStatus().getCode();
             if (OK == status) {
                 // should not happen but just to be safe...
-                return defaultFunction.get();
+                return callToExecute.getDefaultFunction().get();
             } else if (RECOVERABLE_EXCEPTIONS.contains(status)) {
                 // network errors => recoverable => propagate error so OpenNMS can try later again.
                 throw new StorageException(String.format("Network problem %s", status), ex);
             } else {
                 // all other errors: we can't fix them => log and forget...
                 LOG.warn("An error happened during the RPC call: {}", status, ex);
-                return defaultFunction.get();
+                return callToExecute.getDefaultFunction().get();
             }
+        } finally {
+            cloudLog.log(startTime, System.currentTimeMillis(), callToExecute.getMethodDescriptor(), status);
```

Comment:
Do we have any usage case that will log the current time and call the cloudLog.log later? If not I suggest skip the endTime parameter and let the log() handle it by itself.

Reply:
makes sense to me. Done.